File: usaco/contests/2020_dec/bronze/Stuck/Stuck.java
import java.io.*;
import java.util.*;

class Cow {
  char dir; //'E' or 'N'
  int x;
  int y;
  int id; //original order
  Cow (char _dir, int _x, int _y, int _id) {
    dir = _dir;
    x = _x;
    y = _y;
    id = _id;
  }
}

class Pair {
  int x;
  int y;
  Pair (int _x, int _y) {
    x = _x;
    y = _y;
  }
}

class Event implements Comparable<Event> {
  int t;
  Pair p;
  int id; //of cow
  Event(int _t, Pair _p, int _id) {
    t = _t;
    p = _p;
    id = _id;
  }

  //Override the compareTo function, 
  //defining an ordering on events (sorted by time)
  //It's not necessary to completely understand this, but the 
  //Collections.sort function requires that the type of the list 
  //implement the Comparable interface,
  //which is done by implementing the compareTo function.
  @Override
  public int compareTo(Event r) {
    return t-r.t;
  }
}

class Stuck {
  public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());
    Cow[] cows = new Cow[n];
    //Read input into an array of cows
    for (int i = 0; i < n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
      char dir = st.nextToken().charAt(0);
      int x = Integer.parseInt(st.nextToken());
      int y = Integer.parseInt(st.nextToken());
      cows[i] = new Cow(dir,x,y,i);
    }
    //Since we can't simulate the process directly (the bounds are too large),
    //let's come up with a list of "events", or special moments in time, that we care about.
    //Each pair of cows may possibly arrive at the same square, at times t1 and t2.
    //If a cow arrives at a square that has already been eaten, we stop it and mark it as inactive.
    ArrayList<Event> events = new ArrayList<Event>();
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        //Two cows intersect iff one is E and the other is N, and they are positioned in a particular way
        if (cows[i].dir == 'E' && cows[j].dir == 'N'
         && cows[i].x < cows[j].x 
         && cows[i].y > cows[j].y) {
         //These cows will both go to the point (cows[j].x,cows[i].y) 
         int inter_x = cows[j].x;
         int inter_y = cows[i].y;
         int dx = inter_x-cows[i].x;
         int dy = inter_y-cows[j].y;
         if (dx == dy) continue; //If both reach at the same time, ignore it
         events.add(new Event(dx,new Pair(inter_x,inter_y),cows[i].id)); //event for East cow
         events.add(new Event(dy,new Pair(inter_x,inter_y),cows[j].id)); //event for North cow
       }
      }
    }
    //Now process the events. 
    //We sort the events list by time, so we are processing them in order
    Collections.sort(events);
    boolean[] alive = new boolean[n];
    Arrays.fill(alive,true);
    int[] ans = new int[n];
    //We maintain an arraylist, which are the squares that have already been eaten
    ArrayList<Pair> eaten = new ArrayList<Pair>();
    for (Event e: events) {
      if (!alive[e.id]) continue; //If the cow's already inactive, skip this event
      if (contains(eaten,e.p)) { //If the square has already been eaten, stop the cow now
        alive[e.id] = false;
        ans[e.id] = e.t;
      }
      else { 
        //Otherwise, this is the first cow to eat the square, so it gets to keep going
        //Add the square to the list of eaten squares and move on
        eaten.add(e.p);
      }
    }

    //output answer
    for (int i = 0; i < n; i++) {
      if (alive[i]) System.out.println("Infinity");
      else System.out.println(ans[i]);
    }
  }

  //helper function to determine whether a pair is present in the list
  //It's better to use a HashSet or dictionary for this,
  //but I wanted to limit the number of additional data structures that would be used
  static boolean contains(ArrayList<Pair> a, Pair p) {
    for (Pair i: a) {
      if (i.x == p.x && i.y == p.y) return true;
    }
    return false;
  }
}

File: usaco/contests/2020_dec/bronze/Stuck/stuck.py
n = int(input())
cows = []
for i in range(n):
    [a,b,c] = input().split()
    cows.append([a,int(b),int(c),i])

# Simple class to represent an event
class Event:
    def __init__(self, time, point, id):
        self.time = time
        self.point = point
        self.id = id

events = []
for i in cows:
    for j in cows:
        if i[0] == 'E' and j[0] == 'N' and i[1] < j[1] and i[2] > j[2]:
            x = j[1]
            y = i[2]
            dx = x-i[1]
            dy = y-j[2]
            if dx != dy: 
                #if both cows arrive at the intersection at the same time,
                #ignore the intersection
                events.append(Event(dx,(x,y),i[3]))
                events.append(Event(dy,(x,y),j[3]))
# Sort the event list based the time
events.sort(key=lambda x:x.time)
dead = [] #list of id's of cows that have stopped
eaten = [] #list of cells that have already been eaten
ans = [0] * n
for e in events:
    if e.id not in dead: #if the cows has stopped, skip it
        if e.point in eaten: #if this point has been eaten already, stop
            ans[e.id] = e.time
            dead.append(e.id)
        else: #add the point to the list and keep going
            eaten.append(e.point)
# output answer
for i in range(n):
    if i not in dead:
        print("Infinity")
    else:
        print(ans[i])
